README.md
# Semantic Maps

[![Build Status](https://secure.travis-ci.org/SemanticMediaWiki/SemanticMaps.png?branch=master)](http://travis-ci.org/SemanticMediaWiki/SemanticMaps)
[![Scrutinizer Quality Score](https://scrutinizer-ci.com/g/SemanticMediaWiki/SemanticMaps/badges/quality-score.png?s=2e92028475bc897409f8dbb40ce897d6cc88240e)](https://scrutinizer-ci.com/g/SemanticMediaWiki/SemanticMaps/)
[![Latest Stable Version](https://poser.pugx.org/mediawiki/semantic-maps/version.png)](https://packagist.org/packages/mediawiki/semantic-maps)
[![Download count](https://poser.pugx.org/mediawiki/semantic-maps/d/total.png)](https://packagist.org/packages/mediawiki/semantic-maps)

Semantic Maps is an extension that adds semantic capabilities to the [Maps extension]
(https://github.com/JeroenDeDauw/Maps). This
includes the ability to add, edit, aggregate and visualize coordinate data stored through
[Semantic MediaWiki](https://semantic-mediawiki.org/).

Since Semantic Maps uses the Maps API, you can use multiple mapping services. These include
Google Maps (with Google Earth support), OpenLayers and Leaflet.

Both Semantic Maps and Maps are based on Semantic Google Maps and Semantic Layers, and are
meant to replace these extensions. Having Semantic MediaWiki and Maps installed is a
prerequisite for the Semantic Maps extension; the code will not work without it.

## Documentation

* [Usage examples](https://www.semantic-mediawiki.org/wiki/Semantic_Maps_examples)
* [Installation instructions](INSTALL.md)
* [Release notes](RELEASE-NOTES.md)

Note that the installation instructions might be out of date. Initiation now is done
the same way as that of the Maps extension.

## Contributing and support

Semantic Maps 3.4.2 is the last release of Semantic Maps as a dedicated extension.
It has been merged into the [Maps extension](https://github.com/JeroenDeDauw/Maps),
where development continues. All features available in Semantic Maps 3.4.2 are
available in Maps 4.0 and later.


RELEASE-NOTES.md
These are the release notes for the Semantic Maps extension.

Note that Semantic Maps 3.4.2 is the last release of Semantic Maps as a dedicated extension. It has been
merged into the [Maps](https://github.com/JeroenDeDauw/Maps) extension, where development continues. All features
available in Semantic Maps 3.4.2 are now available in Maps 4.0 and later!

## Semantic Maps 3.4.2

Released on November 23rd, 2016.

* Updated the documentation to reflect that this extension is now obsolete and should be replaced by Maps 4.0 and later.

## Semantic Maps 3.4.1

Released on September 21st, 2016.

* Fixed regression introduced in 3.4 that caused the result formats to not be available

## Semantic Maps 3.4

Released on September 18th, 2016.

### Compatibility changes

* Dropped support for MediaWiki older than 1.23
* Dropped support for PHP older than 5.5
* Dropped support for Maps older than 3.8

### New features

* Added support for coordinates in SMWs [record datatype](https://www.semantic-mediawiki.org/wiki/Help:Type_Record) (by James Hong Kong)
* Added support for the Maps tracking category (by Peter Grassberger)

## Semantic Maps 3.3

Released on April 2nd, 2016.

* Added `userparam` support for the map result formats (by James Hong Kong)
* Made Google Maps initialization more robust (by Karlpietsch)
* Added missing system messages (by Karsten Hoffmeyer)

## Semantic Maps 3.2

Released on July 2nd, 2015.

* Added Leaflet result format (by Bernhard Krabina)

## Semantic Maps 3.1.3

Released on November 11th, 2014.

* Restored compatibility with SMW 2.0.x (broken in SM 3.1.2)

## Semantic Maps 3.1.2

Released on November 11th, 2014.

* Updated dependencies to latest versions.
* Improved compatibility with Semantic MediaWiki 2.1.x.

## Semantic Maps 3.1.1

Released on September 14th, 2014.

* Prevent installation with incompatible dependencies

## Semantic Maps 3.1

Released on August 5th, 2014.

* Added compatibility with Semantic MediaWiki 2.x
* Added compatibility with Validator 2.x.

## Semantic Maps 3.0.4

Released on April 3rd, 2014.

* Fixed bug in SM_QueryHandler.php that was causing a fatal error in some cases.

## Semantic Maps 3.0.3

Released on March 28th, 2014.

* Fixed centre parameter bug in the form inputs
* Improved installation instructions
* Translation updates
* Ensured compatibility with PHP 5.6

## Semantic Maps 3.0

Released on January 18th, 2014.

#### Compatibility changes

* Changed minimum Maps version from 2.0.x to 3.0.
* Changed minimum Semantic MediaWiki version from 1.8 to 1.9.
* Changed minimum Semantic Forms version to 2.6.1.
* The extension now needs to be installed via Composer.

#### Removed features

* Dropped the OpenLayers form input, as it was no longer working properly.
* Dropped the "forceshow" parameter, as SMW 1.8 and later do not allow for this functionality.

#### New features

* Support for installation via the Composer dependency manager.

#### Bug fixes

* Fixed compatibility with SMWs SQLStore3 (bug 44933, fixed by kvolden)

#### Internal improvements

* Make it possible to load Semantic Maps before Semantic Forms.
* Added various PHPUnit tests.
* Have the tests run on TravisCI and ScrutinizerCI.

## Semantic Maps 2.0.1 (2012-12-13)

* Improved compatibility with SMW 1.8 and 1.9 (including bug 42830)
* Fixed issue in SMQueryHandler (bug 42250)
* Make use of the default parameter in the query printers

## Semantic Maps 2.0 (2012-10-05)

#### Compatibility changes

* Changed minimum PHP version from 5.2 to 5.3.
* Changed minimum MediaWiki version from 1.17 to 1.18.
* Changed minimum Semantic MediaWiki version from 1.7 to 1.8.
* Changed minimum Maps version from 1.0.x to 2.0.
* Changed minimum Validator version from 0.4 to 0.5.
* Removed support for the deprecated Google Maps v2 API.
* Removed support for the now unsorted Yahoo! Maps API and associated geocoding service.

#### New features

* Added support for the visitedicon parameter to the map result formats.
* Many new functionality that got added in Maps 2.0 (see the Maps release notes).

#### Other improvements

* Improved script loading.
* Usage of Validator 0.5.x features.

#### Bug fixes

* Fixed display of maps in multiple instance templates in forms.
* Fixed partial incompatibility with SMW >=1.6.

## Semantic Maps 1.0.5 (2011-11-30)

* Changed minimun Semantic MediaWiki version from 1.6.1 to 1.7.
* Allow properties of type string to be used as title. (bug 32632)
* Fixed error occuring when parsing invalid coordinates.

## Semantic Maps 1.0.4 (2011-10-15)

* Added compatibility with SMW 1.7 and later.
* Removed compatibility with SMW 1.6.1 and earlier.
* Improved default map format handling.

## Semantic Maps 1.0.3 (2011-09-14)

* Added geocodecontrol parameter for the form inputs.

## Semantic Maps 1.0.2 (2011-08-24)

* Matching Maps 1.0.2 release. 

## Semantic Maps 1.0.1 (2011-08-17)

* Fixed handling of the semantic link caption in the geographical coordinate datavalue class. 

## Semantic Maps 1.0 (2011-07-19)

This version branched from Semantic Maps 0.7.x at version 0.7.3.
See the maps release notes for all changes not specific to Semantic Maps.

* Improved map query parameter support in the Special:Ask GUI

* Fixed fatal error occurring when running maintenance/refreshLinks.php.

* Rewrote all the form inputs to jQuery plugins.

* Rewrote the map printers to use the SMQueryHandler class.

* Added geocoding capability to the OpenLayers form input when having a GeoNames API account.

* Added 'update map' button to all form inputs.

* Fixed PostGres support for distance queries.

* Use of the DataItems introduces in Semantic MediaWiki 1.6.

## Semantic Maps 0.7.3 (2010-11-30) ##

##### new features #####

* Added KML result format that allows the export of geographical data queried via SMW in the form of .kml files.

## Semantic Maps 0.7.2 (2010-10-28) ##

##### Bug fixes #####

* Fixed small layout issue in query pop-ups when property names are blank.


src/queryprinters/SM_MapPrinter.php
<?php

use Maps\Elements\Location;
use Maps\Element;
use Maps\Elements\BaseElement;
use ParamProcessor\ParamDefinition;

/**
 * Query printer for maps. Is invoked via SMMapper.
 * Can be overridden per service to have custom output.
 *
 * @ingroup SemanticMaps
 *
 * @licence GNU GPL v2+
 * @author Jeroen De Dauw < jeroendedauw@gmail.com >
 * @author Peter Grassberger < petertheone@gmail.com >
 */
class SMMapPrinter extends SMW\ResultPrinter {

	private static $services = [];

	/**
	 * @since 3.4
	 * FIXME: this is a temporary hack that should be replaced when SMW allows for dependency
	 * injection in query printers.
	 *
	 * @param MapsMappingService $service
	 */
	public static function registerService( MapsMappingService $service ) {
		self::$services[$service->getName()] = $service;
	}

	public static function registerDefaultService( $serviceName ) {
		self::$services['map'] = self::$services[$serviceName];
	}
	
	/**
	 * @var MapsMappingService
	 */
	private $service;
	
	/**
	 * @var string|boolean
	 */
	protected $fatalErrorMsg = false;
	
	/**
	 * @param string $format
	 * @param bool $inline
	 */
	public function __construct( $format, $inline = true ) {
		$this->service = self::$services[$format];
		
		parent::__construct( $format, $inline );
	}

	/**
	 * Returns an array containing the parameter info.
	 * 
	 * @since 1.0
	 * 
	 * @return array
	 */
	private function getParameterInfo() {
		global $smgQPShowTitle, $smgQPTemplate, $smgQPHideNamespace;
		
		$params = ParamDefinition::getCleanDefinitions( MapsMapper::getCommonParameters() );

		$this->service->addParameterInfo( $params );

		$params['staticlocations'] = [
			'type' => 'mapslocation',
			'aliases' => [ 'locations', 'points' ],
			'default' => [],
			'islist' => true,
			'delimiter' => ';',
			'message' => 'semanticmaps-par-staticlocations',
		];

		$params['showtitle'] = [
			'type' => 'boolean',
			'aliases' => 'show title',
			'default' => $smgQPShowTitle,
		];

		$params['hidenamespace'] = [
			'type' => 'boolean',
			'aliases' => 'hide namespace',
			'default' => $smgQPHideNamespace,
		];

		$params['template'] = [
			'default' => $smgQPTemplate,
		];

		$params['userparam'] = [
			'default' => '',
		];

		$params['activeicon'] =  [
			'type' => 'string',
			'default' => '',
		];

		$params['pagelabel'] =  [
			'type' => 'boolean',
			'default' => false,
		];

		$params['ajaxcoordproperty'] = array(
			'default' => '',
		);

		$params['ajaxquery'] = array(
			'default' => '',
			'type' => 'string'
		);

		// Messages:
		// semanticmaps-par-staticlocations, semanticmaps-par-showtitle, semanticmaps-par-hidenamespace,
		// semanticmaps-par-template, semanticmaps-par-userparam, semanticmaps-par-activeicon,
		// semanticmaps-par-pagelabel, semanticmaps-par-ajaxcoordproperty semanticmaps-par-ajaxquery
		foreach ( $params as $name => &$data ) {
			if ( is_array( $data ) && !array_key_exists( 'message', $data ) ) {
				$data['message'] = 'semanticmaps-par-' . $name;
			}
		}

		$params = array_merge( $params, MapsDisplayMap::getCommonMapParams() );
		
		return $params;
	}
	
	/**
	 * Builds up and returns the HTML for the map, with the queried coordinate data on it.
	 *
	 * @param SMWQueryResult $res
	 * @param $outputmode
	 * 
	 * @return array or string
	 */
	public final function getResultText( SMWQueryResult $res, $outputmode ) {
		if ( $this->fatalErrorMsg !== false ) {
			return $this->fatalErrorMsg;
		}

		/**
		 * @var Parser $wgParser
		 */
		global $wgParser;

		if ( $GLOBALS['egMapsEnableCategory'] && $wgParser->getOutput() !== null ) {
			$wgParser->addTrackingCategory( 'maps-tracking-category' );
		}

		$params = $this->params;

		$queryHandler = new SMQueryHandler( $res, $outputmode );
		$queryHandler->setLinkStyle($params['link']);
		$queryHandler->setHeaderStyle($params['headers']);
		$queryHandler->setShowSubject( $params['showtitle'] );
		$queryHandler->setTemplate( $params['template'] );
		$queryHandler->setUserParam( $params['userparam'] );
		$queryHandler->setHideNamespace( $params['hidenamespace'] );
		$queryHandler->setActiveIcon( $params['activeicon'] );

		$this->handleMarkerData( $params, $queryHandler );
		$locationAmount = count( $params['locations'] );

		$params['ajaxquery'] = urlencode( $params['ajaxquery'] );

		$this->service->addHtmlDependencies(
			MapsDisplayMapRenderer::getLayerDependencies( $params['format'], $params )
		);

		if ( $locationAmount > 0 ) {
			// We can only take care of the zoom defaulting here,
			// as not all locations are available in whats passed to Validator.
			if ( $this->fullParams['zoom']->wasSetToDefault() && $locationAmount > 1 ) {
				$params['zoom'] = false;
			}

			$mapName = $this->service->getMapId();

			SMWOutputs::requireHeadItem(
				$mapName,
				$this->service->getDependencyHtml() .
				$configVars = Skin::makeVariablesScript( $this->service->getConfigVariables() )
			);

			foreach ( $this->service->getResourceModules() as $resourceModule ) {
				SMWOutputs::requireResource( $resourceModule );
			}

			if ( array_key_exists( 'source', $params ) ) {
				unset( $params['source'] );
			}

			return $this->getMapHTML( $params, $wgParser, $mapName );
		}
		else {
			return $params['default'];
		}
	}

	/**
	 * Returns the HTML to display the map.
	 *
	 * @since 1.1
	 *
	 * @param array $params
	 * @param Parser $parser
	 * @param string $mapName
	 *
	 * @return string
	 */
	private function getMapHTML( array $params, Parser $parser, $mapName ) {
		return Html::rawElement(
			'div',
			[
				'id' => $mapName,
				'style' => "width: {$params['width']}; height: {$params['height']}; background-color: #cccccc; overflow: hidden;",
				'class' => 'maps-map maps-' . $this->service->getName()
			],
			wfMessage( 'maps-loading-map' )->inContentLanguage()->escaped() .
				Html::element(
					'div',
					[ 'style' => 'display:none', 'class' => 'mapdata' ],
					FormatJson::encode( $this->getJSONObject( $params, $parser ) )
				)
		);
	}

	/**
	 * Returns a PHP object to encode to JSON with the map data.
	 *
	 * @since 1.1
	 *
	 * @param array $params
	 * @param Parser $parser
	 *
	 * @return mixed
	 */
	private function getJSONObject( array $params, Parser $parser ) {
		return $params;
	}
	
	/**
	 * Converts the data in the coordinates parameter to JSON-ready objects.
	 * These get stored in the locations parameter, and the coordinates on gets deleted.
	 * 
	 * @since 1.0
	 * 
	 * @param array &$params
	 * @param SMQueryHandler $queryHandler
	 */
	private function handleMarkerData( array &$params, SMQueryHandler $queryHandler ) {
		if ( is_object( $params['centre'] ) ) {
			$params['centre'] = $params['centre']->getJSONObject();
		}

		$iconUrl = MapsMapper::getFileUrl( $params['icon'] );
		$visitedIconUrl = MapsMapper::getFileUrl( $params['visitedicon'] );

		$params['locations'] = $this->getJsonForStaticLocations(
			$params['staticlocations'],
			$params,
			$iconUrl,
			$visitedIconUrl
		);

		unset( $params['staticlocations'] );

		$this->addShapeData( $queryHandler->getShapes(), $params, $iconUrl, $visitedIconUrl );

		if ( $params['format'] === 'openlayers' ) {
			$params['layers'] = MapsDisplayMapRenderer::evilOpenLayersHack( $params['layers'] );
		}
	}

	private function getJsonForStaticLocations( array $staticLocations, array $params, $iconUrl, $visitedIconUrl ) {
		/**
		 * @var Parser $wgParser
		 */
		global $wgParser;

		$parser = version_compare( $GLOBALS['wgVersion'], '1.18', '<' ) ? $wgParser : clone $wgParser;

		$locationsJson = [];

		foreach ( $staticLocations as $location ) {
			$locationsJson[] = $this->getJsonForStaticLocation(
				$location,
				$params,
				$iconUrl,
				$visitedIconUrl,
				$parser
			);
		}

		return $locationsJson;
	}

	private function getJsonForStaticLocation( Location $location, array $params, $iconUrl, $visitedIconUrl, Parser $parser ) {
		$jsonObj = $location->getJSONObject( $params['title'], $params['label'], $iconUrl, '', '', $visitedIconUrl );

		$jsonObj['title'] = $parser->parse( $jsonObj['title'], $parser->getTitle(), new ParserOptions() )->getText();
		$jsonObj['text'] = $parser->parse( $jsonObj['text'], $parser->getTitle(), new ParserOptions() )->getText();

		$hasTitleAndtext = $jsonObj['title'] !== '' && $jsonObj['text'] !== '';
		$jsonObj['text'] = ( $hasTitleAndtext ? '<b>' . $jsonObj['title'] . '</b><hr />' : $jsonObj['title'] ) . $jsonObj['text'];
		$jsonObj['title'] = strip_tags( $jsonObj['title'] );

		if ( $params['pagelabel'] ) {
			$jsonObj['inlineLabel'] = Linker::link( Title::newFromText( $jsonObj['title'] ) );
		}

		return $jsonObj;
	}

	/**
	 * @param Element[] $queryShapes
	 * @param array $params
	 * @param string $iconUrl
	 * @param string $visitedIconUrl
	 */
	private function addShapeData( array $queryShapes, array &$params, $iconUrl, $visitedIconUrl ) {
		$params['locations'] = array_merge(
			$params['locations'],
			$this->getJsonForLocations(
				$queryShapes['locations'],
				$params,
				$iconUrl,
				$visitedIconUrl
			)
		);

		$params['lines'] = $this->getElementJsonArray( $queryShapes['lines'], $params );
		$params['polygons'] = $this->getElementJsonArray( $queryShapes['polygons'], $params );
	}

	/**
	 * @param Location[] $locations
	 * @param array $params
	 * @param string $iconUrl
	 * @param string $visitedIconUrl
	 *
	 * @return array
	 */
	private function getJsonForLocations( array $locations, array $params, $iconUrl, $visitedIconUrl ) {
		$locationsJson = [];

		foreach ( $locations as $location ) {
			$jsonObj = $location->getJSONObject( $params['title'], $params['label'], $iconUrl, '', '', $visitedIconUrl );

			$jsonObj['title'] = strip_tags( $jsonObj['title'] );

			$locationsJson[] = $jsonObj;
		}

		return $locationsJson;
	}

	/**
	 * @param BaseElement[] $elements
	 * @param array $params
	 *
	 * @return array
	 */
	private function getElementJsonArray( array $elements, array $params ) {
		$elementsJson = [];

		foreach ( $elements as $element ) {
			$jsonObj = $element->getJSONObject( $params['title'], $params['label'] );
			$elementsJson[] = $jsonObj;
		}

		return $elementsJson;
	}

	/**
	 * Returns the internationalized name of the mapping service.
	 * 
	 * @return string
	 */
	public final function getName() {
		return wfMessage( 'maps_' . $this->service->getName() )->text();
	}
	
	/**
	 * Returns a list of parameter information, for usage by Special:Ask and others.
	 * 
	 * @return array
	 */
    public function getParameters() {
        $params = parent::getParameters();
        $paramInfo = $this->getParameterInfo();
        
        // Do not display this as an option, as the format already determines it
        // TODO: this can probably be done cleaner with some changes in Maps
        unset( $paramInfo['mappingservice'] );
        
        $params = array_merge( $params, $paramInfo );

		return $params;
    }
}
